File: 0739-daily-temperatures/solution.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        Stack<Integer> st = new Stack<>();
        int[] ans = new int[n];

        //st=[6,2]
        // Next Greater Element concept - exact pattern
        // start from end 
        for(int i=n-1; i>=0; i--) {//1
            // check if stack is not empty and stack.peek() <= my current element, POP
            while(!st.isEmpty() && temperatures[st.peek()] <= temperatures[i]) {
                st.pop();
            }

            if(st.isEmpty()) {
                ans[i] = 0;
            } else {
                ans[i] = st.peek() - i;//number of elements greater than current elt
            }
            st.push(i);
        }
        return ans;//[0,0,1,1,2,4]
    }
}

File: 0739-daily-temperatures/solution.swift
class Solution {
    func dailyTemperatures(_ temperatures: [Int]) -> [Int] {
        var n = temperatures.count
        var stack = [Int]()
        var ans = Array(repeating: 0, count: n)

        // NGR -> backwards
        for i in stride(from: n-1, through: 0, by: -1) {
            while !stack.isEmpty && temperatures[i] >= temperatures[stack.last!] {
                stack.removeLast()
            }

            if !stack.isEmpty {
                if let last = stack.last {
                    ans[i] = last - i//1
                }
            }
            stack.append(i)//7
        }
        return ans
    }
}
